django/contrib/admin/templatetags/base.py

```diff
@@ -27,5 +28,4 @@ def render(self, context):
             'admin/%s/%s' % (app_label, self.template_name),
             'admin/%s' % (self.template_name,),
         ]
-
```

Comment:
I'm not sure if the `self.filename` assignment above might violate [thread safety](https://docs.djangoproject.com/en/dev/howto/custom-template-tags/#thread-safety-considerations).

Reply:
Hmm. Maybe it would work to pre-cache the template? It's kind of a hack though.

template = context.template.engine.select_template(filenames)
context.render_context[self] = template

Reply:
I really don't know.
It should be fine because it doesn't change every time it is run (or better: it could change with the same opts values). And users cannot really use a custom template_name without rewriting tags and templates.
But threading what happen with you can't know really.

Reply:
I'm trying to pin this down. It's **not** state between calls (like the `cycle` example). 

Rather, I'm viewing the change list from App A, whilst someone else is viewing the change list from App B. 

My thread sets `self.filename` to the `app-a` `app_label` just as their thread goes to instantiate the template in `super().render()`. They get the override template for the wrong app. 
(i.e. is a race condition.) 

It's pretty tight but it's there. (Or so it seems)

I'm trying out @collinanderson's suggestion. 

I can't think see how to test this.

Reply:
@carltongibson I see the potential problem here (oh, I try to never use threads in first place).
The proposed solution is basically what InclusionNode.render does: I like it :+1: